Create calendar-review.md (#1438)
tc39/proposal-temporal: docs/calendar-review.md

date.monthCode; // => M05L
date.month; // => 6
date.day; // => 2
date.inLeapYear; // => true
date.calendar.id; // => hebrew
inTwoMonths = date.add({ months: 4 }); // => 23 Sivan 5779
inTwoMonths.withCalendar('iso8601'); // => 2019-06-26
date.until(inTwoMonths, { largestUnit: 'months' }); // => P4M
```

By definition, creating a bag of fields with year, month, and day requires choosing a calendar system. We elide the ISO calendar if no other calendar system is specified in this situation, and it is the only place Temporal performs calendar elision:

```javascript
Temporal.PlainDate.from({ year: 2019, month: 2, day: 28 });
```

The following Temporal types all carry a calendar: `Temporal.ZonedDateTime`, `Temporal.PlainDateTime`, `Temporal.PlainDate`, `Temporal.PlainYearMonth`, `Temporal.PlainMonthDay`

## The Unexpected Calendar Problem

Integrating calendar awareness into many Temporal operations is essential for a global audience, but it introduces a new problem: if code is written to assume one calendar system, bugs may result if a different calendar system is introduced instead. Introducing unexpected calendars could also be a vector for malicious activity.

> This section describes this problem, which we call the “Unexpected Calendar Problem,” and how to mitigate it in userland Temporal code.

Reviewer: ```suggestion
This section describes this problem, which we call the "Unexpected Calendar Problem," and how to mitigate it in userland Temporal code.
```

### Problem Definition

Below is a canonical example of this problem: a function which tests whether a particular date is the last day of the year. It assumes a calendar with 12 months and whose last month is 31 days long.

```javascript
function isLastDayOfYear(date) {
  // Note: This is a BAD example!
  return date.month === 12 && date.day === 31;
}

// works for ISO calendar
isLastDayOfYear(Temporal.PlainDate.from('2019-12-31')); // => true
isLastDayOfYear(Temporal.PlainDate.from('2020-01-01')); // => false

// fails for some non-ISO calendars
hebrewNewYearsEve = Temporal.PlainDate.from(
  { year: 5780, monthCode: 'M12', day: 29, calendar: 'hebrew' });
isLastDayOfYear(hebrewNewYearsEve); // => expected: true, actual: false
```

### Mitigating the Unexpected Calendar Problem

There are two ways to avoid the Unexpected Calendar Problem: the preferred solution of writing calendar-safe code, or a fallback option to validate or coerce inputs to a known calendar.

#### Writing Calendar-Safe Code

> The best way to mitigate the Unexpected Calendar Problem is to write “calendar-safe” code that works for all built-in calendars. For example, the function above can easily be written to be calendar-safe by using the `monthsInYear` and `daysInMonth` properties instead of hard-coding ISO constants. Developers who follow documented [best practices for writing calendar-safe Temporal code](https://tc39.es/proposal-temporal/docs/calendar.html#writing-cross-calendar-code) will be able to use the same code for all built-in calendars.

Reviewer: ```suggestion
The best way to mitigate the Unexpected Calendar Problem is to write "calendar-safe" code that works for all built-in calendars. For example, the function above can easily be written to be calendar-safe by using the `monthsInYear` and `daysInMonth` properties instead of hard-coding ISO constants. Developers who follow documented [best practices for writing calendar-safe Temporal code](https://tc39.es/proposal-temporal/docs/calendar.html#writing-cross-calendar-code) will be able to use the same code for all built-in calendars.
```

Beyond documentation, the Temporal API itself has been designed to make it easier to write calendar-safe code. For example:

*   All built-in calendars (including ISO) use the same set of date fields: `year`, `month`, `day`, and `monthCode`. These fields act the same way in all built-in calendars.
*   The `month` property is a continuous (no gaps), 1-based index of the month in the current year. This aligns with ISO behavior and avoids unexpected behavior from lunisolar calendars where each year may have a different set of months.  A separate `monthCode` string field is used for a year-independent month identifier.

> *   The `year` property is a signed value that’s relative to a specific calendar era. This matches ISO calendar behavior and prevents bugs from, for example, eras like B.C. where years count backwards.  Separate `eraYear` and `era` fields are used for measuring years within the context of a specific calendar era.

Reviewer: Avoiding the repetition of "specific era" may make this section clearer.

```suggestion
*   The `year` property is a signed value that's relative to a specific "default era" for each calendar. This matches ISO calendar behavior and prevents bugs from, for example, eras like B.C. where years count backwards.  Separate `eraYear` and `era` fields are used for measuring years within the context of other eras.
```

*   Differences between calendars are exposed via convenience properties available in every calendar, e.g. `monthsInYear`, `daysInMonth`, `inLeapYear`, etc. Userland code can completely avoid the use of calendar-specific constants (e.g. 12 months per year) by using these properties. 

> *   There is never an implicit default calendar system. Developers must explicitly decide whether they’re using the ISO calendar or another calendar. For example, shortcut methods on the `Temporal.now` object are named with an ISO prefix to avoid ambiguity, such as `Temporal.now.zonedDateTimeISO()`.

Reviewer: ```suggestion
*   There is never an implicit default calendar system. Developers must explicitly decide whether they're using the ISO calendar or another calendar. For example, shortcut methods on the `Temporal.now` object are named with an ISO prefix to avoid ambiguity, such as `Temporal.now.zonedDateTimeISO()`.
```

> *   Temporal never infers a calendar from the user’s environment.  Introducing a calendar system into a program requires a conscious, opt-in decision on the part of the developer of the app, library, or input source.

Reviewer: ```suggestion
*   Temporal never infers a calendar from the user's environment.  Introducing a calendar system into a program requires a conscious, opt-in decision on the part of the developer of the app, library, or input source.
```

#### Validating or Coercing Inputs

> Even though Temporal makes it straightforward to write calendar-safe code, doing so still requires work from developers. Some developers won’t do this work. Sometimes this will be because they don’t know about or don’t understand non-ISO calendars. In other cases, developers won’t bother to follow best practices because they expect to only be dealing with ISO-calendar data. Even if a developer conscientiously follows best practices in their own code, most ECMAScript apps have many library dependencies that may not do so. Or they may have to interoperate with existing code that cannot be changed due to time constraints.

Reviewer: ```suggestion
Even though Temporal makes it straightforward to write calendar-safe code, doing so still requires work from developers. Some developers won't do this work. Sometimes this will be because they don't know about or don't understand non-ISO calendars. In other cases, developers won't bother to follow best practices because they expect to only be dealing with ISO-calendar data. Even if a developer conscientiously follows best practices in their own code, most ECMAScript apps have many library dependencies that may not do so. Or they may have to interoperate with existing code that cannot be changed due to time constraints.
```

In these situations, the developer should ensure that inputs to that code are either (a) validated to ensure the input is in the expected calendar system, or (b) coerced to that system. Whether to validate or coerce depends on the use case.

To coerce:

```javascript
date = date.withCalendar('iso8601');
```

To validate:

```javascript
if (date.calendar.id !== 'iso8601') throw new Error ('invalid calendar');
```

> Note that “inputs” are not only external data. Inputs could include values returned from library dependencies, e.g. date pickers. Unless the library or input source guarantees a certain calendar system, defensive code should assume Temporal objects (or strings or objects that will be converted into a Temporal object) to be “external” and subject to validation or coercion to the ISO calendar if the code that will use that data is not calendar-safe.

Reviewer: ```suggestion
Note that "inputs" are not only external data. Inputs could include values returned from library dependencies, e.g. date pickers. Unless the library or input source guarantees a certain calendar system, defensive code should assume Temporal objects (or strings or objects that will be converted into a Temporal object) to be "external" and subject to validation or coercion to the ISO calendar if the code that will use that data is not calendar-safe.
```

*   Many methods (`add`, `subtract`, `round`, `until`, `since`, `with`, maybe `toPlainMonthDay`) would require a calendar option to be able to use calendar-specific months, days, or years. Example: \
`date.add({months: 2}, {calendar: 'chinese'})`
*   Chaining multiple operations (e.g. `.add({months: 2}).with({day: 1})`) gets less ergonomic and more bug-prone because the same calendar value must be used in all chained calls. 
*   The `Temporal.PlainMonthDay` type would become challenging to reason about without a calendar attached because month/day values like birthdays and holidays are inherently calendar-specific and have no ISO counterpart. This is especially true in lunisolar calendars like Hebrew and Chinese where ordinal month numbers vary from year to year for the same month. (Current Temporal uses a string <code>[monthCode](https://tc39.es/proposal-temporal/docs/plaindate.html#monthCode)</code> field to describe year-independent months.) For example, a birthday on 12 Adar I in the Hebrew calendar is currently modelled as \
<code>  Temporal.PlainMonthDay.from( \
    { monthCode: '05L', day: 12, calendar: 'hebrew' } \
  ); \

> </code>It’s not clear how we’d model holidays if the calendar isn’t stored in the instance.

Reviewer: ```suggestion
</code>

It's not clear how we'd model holidays if the calendar isn't stored in the instance.
```

> *   Code that wants to return a date along with its calendar can no longer return a single Temporal instance. Instead, it must return a compound object or serialized string. For example, a date picker that returned a date in the user’s preferred calendar could no longer return a <code>Temporal.PlainDate</code>. Instead, it’d have to return a tuple or compound object like <code>{date, calendar}</code>.  Passing around compound objects (or extra calendar parameters) would complicate all calendar-using Temporal code.

Reviewer: ```suggestion
*   Code that wants to return a date along with its calendar can no longer return a single Temporal instance. Instead, it must return a compound object or serialized string. For example, a date picker that returned a date in the user's preferred calendar could no longer return a <code>Temporal.PlainDate</code>. Instead, it'd have to return a tuple or compound object like <code>{date, calendar}</code>.  Passing around compound objects (or extra calendar parameters) would complicate all calendar-using Temporal code.
```

The Temporal champions believe that the ergonomic challenges of this alternative are worse than the work required to mitigate the Unexpected Calendar Problem. 

### Alternative 2: Separate ISO and Calendar Properties

> It would be possible to retain the calendar in Temporal objects’ data model without being vulnerable to the Unexpected Calendar Problem, but doing so would involve a significant increase in Temporal surface area. Below is one way it could be done. 

Reviewer: ```suggestion
It would be possible to retain the calendar in Temporal objects' data model without being vulnerable to the Unexpected Calendar Problem, but doing so would involve a significant increase in Temporal surface area. Below is one way it could be done. 
```

    *   Instances using a non-ISO calendar would only support calendar fields: `calendarDays`, `calendarWeeks`, `calendarMonths`, and `calendarYears`. 
    *   String parsing and serialization would use calendar annotations like other Temporal types, e.g. `'P2D[u-ca-chinese]'`. Strings without an annotation would assume the ISO calendar.
    *   Question: should `date.add({calendarMonths: 1})` use `date.calendar`? Or should `date.add({calendarMonths: 1, calendar: 'chinese'})` be required even if `date.calendar==='chinese'`?
    *   `until` and `since` are now ambiguous, so they’d need a way to indicate whether the output should be an ISO duration or a non-ISO one. Options include:
        *   Split each method into "iso" and "calendar" variants, e.g. `isoUntil` vs. `calendarUntil` or `untilISO` vs. `untilCalendar`
        *   Always return both flavors, e.g. `foo.until(bar).isoDuration` and `foo.until(bar).calendarDuration`.
        *   Use an option, e.g. `{ resultFields: 'iso' | 'calendar' }`. FWIW, I'm not sure that this option would work well with TS in cases where options are not specified as a literal in the API call.

>     *   There would NOT be a `withCalendar` method on durations, because durations don’t have the ability to convert between calendar and ISO data. Instead, they would either have ISO fields or calendar fields, but not both.

Reviewer: ```suggestion
    *   There would NOT be a `withCalendar` method on durations, because durations don't have the ability to convert between calendar and ISO data. Instead, they would either have ISO fields or calendar fields, but not both.
```

> After considering this and similar options, the champions decided that the additional complexity wasn’t worth it. In particular, we were concerned that the dual-property approach would be more confusing and would result in more bugs than teaching users to validate or coerce inputs. 

Reviewer: ```suggestion
After considering this and similar options, the champions decided that the additional complexity wasn't worth it. In particular, we were concerned that the dual-property approach would be more confusing and would result in more bugs than teaching users to validate or coerce inputs. 
```

# Supporting Calendar Systems in Temporal

> Authors: Temporal champions [justingrant](https://twitter.com/justingrantjg) and [sffc](https://twitter.com/_sffc)

Reviewer: One more note: do we need to call out authors here?  We don't do that anywhere else in Temporal docs, so not sure there's a need for it here.

Author: Removed

*   令和3年3月4日 == March 4, 3 Reiwa (Japanese)

Non-Gregorian calendar systems are used in day-to-day life in several countries, particularly those with weaker cultural and economic ties to Gregorian-dominant Western countries, as well as in religious, cultural, and academic applications.  For example, several countries including Japan, China, Iran, Afghanistan, Saudi Arabia, and Taiwan use non-Gregorian calendars for at least some official government purposes. Beyond those official uses, a majority of the world's population lives in countries that use non-Gregorian calendars to determine the dates of religious or cultural holidays. 

### Calendars in Business Logic (Not Only String Localization)

Calendar systems play an important role in localization (l10n).  They enable a human-friendly display representation of timestamps, dates, and times.

However, calendar systems also play an important role in the business logic of an application.  Applications involving dates intended for human consumption must consider the calendar system when performing arithmetic or any other business-logic operation that uses months and years.  Examples of calendar-sensitive operations include:

> *   What is the current day plus one month?

Reviewer 2: Maybe to sound slightly more natural, "What is the date one month from today?"

// Create a Hebrew date from a bag of fields
Temporal.PlainDate.from({
  year: 5779,
  monthCode: 'M05L',
  day: 23,
  calendar: 'hebrew'
});

// Create an ISO date and convert it to Hebrew
Temporal.PlainDate.from('2019-02-28').withCalendar('hebrew');
```

Internally, the day is always represented in the ISO calendar, meaning that if the calendar annotation gets dropped, the day remains the same.  This is true for both the internal slots of Temporal types and the string representation.  Thus, the following two lines also result in the same 23 Adar I 5779 instance:

```javascript
// Parse from an annotated ISO string (IETF RFC pending)
Temporal.PlainDate.from('2019-02-28[u-ca=hebrew]');

// Feed the internal data model via the constructor
new Temporal.PlainDate(2019, 2, 28, 'hebrew');
```

Operations on Temporal instances interpret all month, day, and year values in context of the calendar system. For example: 

```javascript
date = Temporal.PlainDate.from('2019-02-28[u-ca=hebrew]');
date.with({ day: 1 }); // => 1 Adar I 5779
date.year; // => 5779
date.monthCode; // => M05L
date.month; // => 6
date.day; // => 2
date.inLeapYear; // => true
date.calendar.id; // => hebrew

> inTwoMonths = date.add({ months: 4 }); // => 23 Sivan 5779
> inTwoMonths.withCalendar('iso8601'); // => 2019-06-26
> date.until(inTwoMonths, { largestUnit: 'months' }); // => P4M

Reviewer 2: ```suggestion
inFourMonths = date.add({ months: 4 }); // => 23 Sivan 5779
inFourMonths.withCalendar('iso8601'); // => 2019-06-26
date.until(inFourMonths, { largestUnit: 'months' }); // => P4M
```

*   Many methods (`add`, `subtract`, `round`, `until`, `since`, `with`, maybe `toPlainMonthDay`) would require a calendar option to be able to use calendar-specific months, days, or years. Example: \
`date.add({months: 2}, {calendar: 'chinese'})`
*   Chaining multiple operations (e.g. `.add({months: 2}).with({day: 1})`) gets less ergonomic and more bug-prone because the same calendar value must be used in all chained calls. 
*   The `Temporal.PlainMonthDay` type would become challenging to reason about without a calendar attached because month/day values like birthdays and holidays are inherently calendar-specific and have no ISO counterpart. This is especially true in lunisolar calendars like Hebrew and Chinese where ordinal month numbers vary from year to year for the same month. (Current Temporal uses a string <code>[monthCode](https://tc39.es/proposal-temporal/docs/plaindate.html#monthCode)</code> field to describe year-independent months.) For example, a birthday on 12 Adar I in the Hebrew calendar is currently modelled as \
<code>  Temporal.PlainMonthDay.from( \

>     { monthCode: '05L', day: 12, calendar: 'hebrew' } \

Reviewer 2: ```suggestion
    { monthCode: 'M05L', day: 12, calendar: 'hebrew' } \
```

    *   The number and types of `calendar*` fields are completely up to the calendar implementation.  Built-in calendars will all offer `calendarDay`, `calendarMonth`, `calendarMonthCode`, `calendarYear`, and sometimes `calendarEra` and `calendarEraYear` too.
    *   Temporal APIs that accept property bag parameters will accept either calendar fields or ISO fields, BUT NOT BOTH. `date.with({calendarYear: 5780, isoMonth: 12}) `should throw. BTW, this may be enforceable via TS and ESLint for a nice design-time DX.
    *   Trying to read `calendar*` fields would throw if the calendar is ISO because there is no human calendar defined.
    *   Non-field convenience properties (e.g. `dayOfYear`, `inLeapYear`, etc.) would also be duplicated. The non-`calendar` variants of these fields would report the ISO values.
*   `Temporal.Duration` would gain a `calendar` field.  
    *   Instances using the ISO calendar would only support ISO fields: `days`, `weeks`, `months`, and `years`. 
    *   Instances using a non-ISO calendar would only support calendar fields: `calendarDays`, `calendarWeeks`, `calendarMonths`, and `calendarYears`. 

>     *   String parsing and serialization would use calendar annotations like other Temporal types, e.g. `'P2D[u-ca-chinese]'`. Strings without an annotation would assume the ISO calendar.

Reviewer 2: ```suggestion
    *   String parsing and serialization would use calendar annotations like other Temporal types, e.g. `'P2D[u-ca=chinese]'`. Strings without an annotation would assume the ISO calendar.
```

>     *   Question: should `date.add({calendarMonths: 1})` use `date.calendar`? Or should `date.add({calendarMonths: 1, calendar: 'chinese'})` be required even if `date.calendar==='chinese'`?

Reviewer 2: This section kind of mixes up the modifications that would need to be made in order to take this approach, with the consequences of making those modifications. FWIW, I think the document is fine this way, so no need to block merging on this, but it does read slightly confusingly.

Author: I removed it.  I think there are enough other bullets in this section.  (OK @justingrant?)
